### PR TITLE
Make the catalog more efficient

### PR DESCRIPTION
**Pull request by pdevine**
This change removes the sort() from the Repositories() function since
we're now guaranteed to have a lexigraphically sorted walk.  Also added
is the new ErrFinishedWalk which signals the walk function to stop feeding
files and directories since no more can be processed.

Signed-off-by: Patrick Devine <patrick.devine@docker.com>

### CODE REVIEW

**Comment by pdevine** [comment]
I'm not sure why janky is failing here, but it looks unrelated to the PR.

File: registry/storage/walk.go [hunk 1]
```diff
@@ -38,9 +43,16 @@ func Walk(ctx context.Context, driver storageDriver.StorageDriver, from string,
...
-			return err
+
+		skipDir := false
+		if err != nil {
+			if err == ErrFinishedWalk {
+				break
+			} else if err == ErrSkipDir {
+				skipDir = true
+			} else {
+				return err
```
Comment (stevvooe) [inline registry/storage/walk.go]: This condition should be sufficient to complete the walk. A `break` here will cause a `nil` return, which would work like a skip dir.

**Comment by stevvooe** [comment]
@pdevine Don't worry about janky. We are still trying to get those stabilized.

File: registry/storage/walk.go (hunk 1 above)
Comment (pdevine) [inline registry/storage/walk.go]: The nil return should pass back a nil and terminate the Walk(), which isn't quite the same as skipDir here, but your point is valid;  I can just modify the called walk function and have it pass back any old error to get the same result.

**Comment by RichardScothern** [comment]
LGTM

**Comment by stevvooe** [comment]
LGTM

**Pull request merged**